linux/DtaDevLinuxSata.cpp

```diff
@@ -413,6 +525,56 @@ static void safecopy(uint8_t * dst, size_t dstsize, uint8_t * src, size_t srcsiz
     if (size < dstsize) memset(dst+size, '\0', dstsize-size);
 }
 
+static void get_serial_from_vpd(int fd, uint8_t *resp_buf)
+{
+    sg_io_hdr_t sg;
+    uint8_t sense[18];
+    uint8_t cdb[sizeof(CScsiCmdInquiry)];
+    uint8_t *buffer = (uint8_t *) aligned_alloc(IO_BUFFER_ALIGNMENT, MIN_BUFFER_LENGTH);
+
+    memset(&cdb, 0, sizeof (cdb));
+    memset(&sense, 0, sizeof (sense));
+    memset(&sg, 0, sizeof (sg));
+
+    // fill out SCSI command
+    auto p = (CScsiCmdInquiry *) cdb;
+    p->m_Opcode = p->OPCODE;
+    p->m_EVPD = 0x1;
+    p->m_PageCode = 0x80;
+    p->m_AllocationLength = htons(sizeof(CScsiCmdInquiry_VPD80));
+
+    // fill out SCSI Generic structure
+    sg.interface_id = 'S';
+    sg.dxfer_direction = SG_DXFER_FROM_DEV;
+    sg.cmd_len = sizeof (cdb);
+    sg.mx_sb_len = sizeof (sense);
+    sg.iovec_count = 0;
+    sg.dxfer_len = MIN_BUFFER_LENGTH;
+    sg.dxferp = buffer;
+    sg.cmdp = cdb;
+    sg.sbp = sense;
+    sg.timeout = 60000;
+    sg.flags = 0;
+    sg.pack_id = 0;
+    sg.usr_ptr = NULL;
+
+    // execute I/O
+    if (ioctl(fd, SG_IO, &sg) < 0) {
+        LOG(D4) << "SG_IO ioctl error: " << strerror(errno);
+        free(buffer);
+        return;
+    }
+    if ((sg.info & SG_INFO_OK_MASK) != SG_INFO_OK) {
+        free(buffer);
+        return;
+    }
+
+    auto resp = (CScsiCmdInquiry_VPD80 *) buffer;
+    memcpy(resp_buf, resp->m_Serial, htons(resp->m_PageLength));
+    scsi_truncate_spaces(resp_buf, 20);
```

Comment:
SCSI serial number is not limited to 20 bytes IIRC.  You might get buffer overflow here.

Reply:
Thanks, updated to `safecopy`!